parsl/executors/high_throughput/probe.py

```diff
@@ -1,62 +1,70 @@
+from parsl import curvezmq
 
 logger = logging.getLogger(__name__)
 
 
-def probe_addresses(addresses, port, timeout=120):
+def probe_addresses(
+    zmq_context: curvezmq.ClientContext,
+    addresses: str,
+    timeout_ms: int = 120_000,
+    identity: Optional[bytes] = None,
+):
     """
-    Parameters
-    ----------
-
-    addresses: [string]
-        List of addresses as strings
-    port: int
-        Port on the interchange
-    timeout: int
-        Timeout in seconds
-
-    Returns
-    -------
-    None or string address
+    Given a single-line CSV list of addresses, return the first proven valid address.
+
+    This function will connect to each address in ``addresses`` (a comma-separated
+    list of URLs) and attempt to send a CONNECTION_PROBE packet.  Returns the first
+    address that receives a response.
+
+    If no address receives a response within the ``timeout_ms`` (specified in
+    milliseconds), then raise ``ConnectionError``.
+
+    :param zmq_context: A ZMQ Context; the call-site may provide an encrypted ZMQ
+        context for assurance that the returned address is the expected and correct
+        endpoint
+    :param addresses: a comma-separated string of addresses to attempt.  Example:
+        ``tcp://127.0.0.1:1234,tcp://[3812::03aa]:5678``
+    :param timeout_ms: how long to wait for a response from the probes.  The probes
+        are initiated and await concurrently, so this timeout will be the total wall
+        time in the worst case of "no addresses are valid."
+    :param identity: a ZMQ connection identity; used for logging connection probes
+        at the interchange
+    :raises: ``ConnectionError`` if no addresses are determined valid
+    :returns: a single address, the first one that received a response
     """
-    context = zmq.Context()
-    addr_map = {}
-    for addr in addresses:
-        socket = context.socket(zmq.DEALER)
-        socket.setsockopt(zmq.LINGER, 0)
-        socket.setsockopt(zmq.IPV6, True)
-        url = tcp_url(addr, port)
-        logger.debug("Trying to connect back on {}".format(url))
-        socket.connect(url)
-        addr_map[addr] = {'sock': socket,
-                          'mon_sock': socket.get_monitor_socket(events=zmq.EVENT_CONNECTED)}
-
-    start_t = time.time()
-
-    first_connected = None
-    while time.time() < start_t + timeout and not first_connected:
-        for addr in addr_map:
-            try:
-                recv_monitor_message(addr_map[addr]['mon_sock'], zmq.NOBLOCK)
-                first_connected = addr
-                logger.info("Connected to interchange on {}".format(first_connected))
-                break
-            except zmq.Again:
-                pass
-            time.sleep(0.01)
-
-    for addr in addr_map:
-        addr_map[addr]['sock'].close()
-
-    return first_connected
+    if not addresses:
+        raise ValueError("No address to probe!")
+
+    sock_map = {}
+    urls = addresses.split(",")
```

Comment:
Since we split the CSV here, why not accept a list of strings for the argument?

Reply:
Because this PR is making way for an upcoming change that will no longer pass around the `host` and `port` as two separate entities.  Instead, the executor will start the managers with a precise list of full TCP URLs to attempt.  The core detail of this function is to return a known-good address ... so could move that to outside of the probe function, but it's just as easy to keep the string manipulation isolated here.